build-tools/xaprepare/xaprepare/Steps/Step_DetermineAzurePipelinesTestJobs.cs

```diff
@@ -0,0 +1,98 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Xamarin.Android.Prepare
+{
+	class Step_DetermineAzurePipelinesTestJobs : Step
+	{
+		public Step_DetermineAzurePipelinesTestJobs ()
+			: base ("Determine Azure Pipelines test jobs to run from a merge commit")
+		{}
+
+		protected override async Task<bool> Execute (Context context)
+		{
+			Log.StatusLine ("Determining test jobs to run...");
+
+			string commitRevision = Environment.GetEnvironmentVariable ("BUILD_SOURCEVERSION");
+			string commitMessage = Environment.GetEnvironmentVariable ("BUILD_SOURCEVERSIONMESSAGE");
+			if (string.IsNullOrEmpty (commitRevision) || string.IsNullOrEmpty (commitMessage)) {
+				Log.ErrorLine ("One or more source version variable values were empty:");
+				Log.ErrorLine ($"BUILD_SOURCEVERSION='{commitRevision}' BUILD_SOURCEVERSIONMESSAGE='{commitMessage}'.");
+				return false;
+			}
+
+			// Assume we're building a merge commit as part of an Azure Pipelines PR build. Otherwise, this step will fail.
+			//  Example: Merge 0b66502c8b9f33cbb8d21b2dab7c100629aec081 into 0bef8aa5cd74d83d77c4e2b3f63975a0deb804b3
+			var commitMessagePieces = commitMessage.Split (new string [] { " " }, StringSplitOptions.RemoveEmptyEntries);
+			if (string.IsNullOrEmpty (commitMessagePieces [3])) {
+				Log.ErrorLine ($"Unable to parse merge commit message from: '{commitMessage}'.");
+				return false;
+			}
+
+			var git = new GitRunner (context);
+			var filesChanged = await git.RunCommandForOutputAsync (BuildPaths.XamarinAndroidSourceRoot, "diff", "--name-only", commitRevision, commitMessagePieces [3]);
+			if (filesChanged == null || filesChanged.Count < 1) {
+				Log.ErrorLine ($"Unable to determine if any files were changed in this PR.");
+				return false;
+			}
+
+			var testAreas = new List<string> ();
+			foreach (string file in filesChanged) {
+				Log.InfoLine ($"Detected change in file: '{file}'.");
+				// Compare files changed against common areas requiring additional test scope.
+				// MSBuild: Runs all legacy and One .NET Xamarin.Android.Build.Task tests
+				// MSBuildDevice: Runs all MSBuildDeviceIntegration tests
+				// BCL: Runs BCL tests on emulator
+				// TimeZone: Runs timezone unit tests on emulator
+				// Designer: Runs designer integration tests
+				if (file == ".external") {
```

Comment:
Aside: what runs the on-device "apk" tests? Is that a separate area?

Might need a semi-exhaustive listing of `src/*`?

   * `src/Microsoft.Android.Sdk.ILLink`: OneDotNet tests + apk tests
   * `src/Mono.Android`: apk tests, Designer
   * `src/Mono.Android.Export`: apk tests
   * `src/OpenTK-1.0`: apk tests
   * `src/System.EnterpriseServices`: ignore
   * `src/Xamarin.Android.Build.Tasks`: MSBuild, MSBuildDevice, Designer
   * `src/Xamarin.Android.NUnitLite`: apk tests
   * `src/Xamarin.Android.NamingCustomAttributes`: ignore
   * `src/Xamarin.Android.Tools.Aidl`: MSBuild
   * `src/Xamarin.Android.Tools.JavadocImporter`: MSBuild
   * `src/aapt2`: MSBuild
   * `src/apksigner`: MSBuild? MSBuildDevice?
   * `src/bundletool`: MSBuild? MSBuildDevice?
   * `src/java-runtime`: apk tests
   * `src/manifestmerger`: MSBuild? MSBuildDevice?
   * `src/monodroid`: apk tests
   * `src/proguard`: MSBuild? MSBuildDevice? apk tests?
   * `src/r8`: MSBuild? MSBuildDevice? apk tests?
   * `src/sqlite-xamarin`: apk tests

Reply:
The the on-device "apk" tests will continue to run on every PR build as they are part of the smoke test stage. That job executes in under one hour and I felt that we didn't need to try to cut scope there at this time. As a result, we won't need to associate any source changes with [that particular test job](https://github.com/xamarin/xamarin-android/blob/05620bb1dedf499aeb80ee2e1ea4d52dd1eef64e/build-tools/automation/azure-pipelines.yaml#L399-L658).

I think we could run extra MSBuild test jobs based on other `src` changes as suggested, @jonathanpeppers and/or @dellis1972 what are your thoughts on the source paths above?

Reply:
The ones above look right to me. `proguard` doesn't need the MSBuild One .NET tests.

Maybe could also add:
* `src-ThirdParty/android-platform-tools-base` - MSBuild, MSBuildDevice
* `src-ThirdParty/bazel` - MSBuildDevice
* `tests/MSBuildDeviceIntegration` - MSBuildDevice
* `tests/msbuild-times-reference` - MSBuildDevice

Reply:
`src/java-runtime` will also need MSBuild tests I think and MSBuildDevice. This folder contains the release/debug runtimes. I don't think the apk tests will test debug/fastdev at all, so we might not get good coverage just running those tests.

Reply:
same for monodroid changes.

Reply:
I think I've addressed all of these source areas.